docs/ha-considerations.md

```diff
@@ -0,0 +1,337 @@
+frontend apiserver
+    bind *:${APISERVER_DEST_PORT}
+    mode tcp
+    option tcplog
+    default_backend apiserver
+
+#---------------------------------------------------------------------
+# round robin balancing for apiserver
+#---------------------------------------------------------------------
+backend apiserver
+    option httpchk GET /healthz
+    http-check expect status 200
+    mode tcp
+    option ssl-hello-chk
+    balance     roundrobin
+        server ${HOST1_ID} ${HOST1_ADDRESS}:${APISERVER_SRC_PORT} check
+        # [...]
+```
+Again, there are some placeholders in `bash` variable style to expand:
+- `${APISERVER_DEST_PORT}` the port through which Kubernetes will talk to the API Server.
+- `${APISERVER_SRC_PORT}` the port used by the API Server instances
+- `${HOST1_ID}` a symbolic name for the first load-balanced API Server host
+- `${HOST1_ADDRESS}` a resolvable address (DNS name, IP address) for the first load-balanced API Server host
+- additional `server` lines, one for each load-balanced API Server host
+
+### Option 1: Run the services on the operating system
+
+In order to run the two services on the operating system, the respective distribution's package manager can be used to install the software. This can make sense if they will be running on dedicated hosts not part of the Kubernetes cluster.
+
+On a RedHat-style Linux distribution this will most likely look like this:
+```
+# yum install keepalived haproxy
+```
+Having now installed the abovementioned configuration, the services can be enabled and started. On a recent RedHat-based system, `systemd` will be used for this:
+```
+# systemctl enable haproxy --now
+# systemctl enable keepalived --now
+```
+With the services up, now the Kubernetes cluster can be bootstrapped using `kubeadm init` (see [below](#bootstrap-the-cluster)).
+
+### Option 2: Run the services as static pods
+
+If `keepalived` and `haproxy` will be running on the control plane nodes they can be configured to run as static pods. All that is necessary here is placing respective manifest files in the _/etc/kubernetes/manifests_ directory before bootstrapping the cluster. During the bootstrap process, `kubelet` will bring the processes up, so that the cluster can use them while starting. This is an elegant solution, in particular with the setup described under [Stacked control plane and etcd nodes](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/high-availability/#stacked-control-plane-and-etcd-nodes), where the `etcd` cluster, too, will be running as static pods on the control plane nodes.
+
+For this setup, two manifest files need to be created in _/etc/kubernetes/manifests_ (create the directory first).
+
+The manifest for `keepalived`, _/etc/kubernetes/manifests/keepalived.yaml_:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  creationTimestamp: null
+  name: keepalived
+  namespace: kube-system
+spec:
+  containers:
+  - image: osixia/keepalived:1.3.5-1
```

Comment:
i haven't seen the name osixia before.
do we have a more "official" image maintained at docker hub?

Reply:
> i haven't seen the name osixia before.
> do we have a more "official" image maintained at docker hub?

I did not find anything. While there is a Dockerfile distributed there isn't anything labeled as "official" or even "certified" on Docker Hub.

Reply:
ok, understood. maybe someone else will have ideas.